architecture-decisions/0004-preservation-fixity.md

```diff
@@ -0,0 +1,111 @@
+        from Google to handle the download, and the function restricted to 100 concurrent
+        workers. If we need more concurrency later we can request a higher
+        inbound traffic quota.
+   1. The Google Cloud Function will send a message to a Status Topic, which
+      will contain a message giving the status of the operation. It looks like
+      the following:
+      ```json
+      {
+        "status": "[SUCCESS/FAILURE]",
+        "resource_id": "[preservation_object_id]",
+        "child_id": "[file_metadata_id]",
+        "child_property": "[property_file_metadata_node_is_stored_in]"
+      }
+      ```
+      - Production Status Topic is configured with the following commands:
+        ```
+        gcloud beta pubsub topics create figgy-production-fixity-status
+        echo '{"bindings":[{"members":["serviceAccount:figgy-preservation-production@pulibrary-figgy-storage-1.iam.gserviceaccount.com"],"role":"roles/pubsub.editor"}],"etag":"ACAB"}' > permissions.json
+        gcloud beta pubsub topics set-iam-policy projects/pulibrary-figgy-storage-1/topics/figgy-production-fixity-status permissions.json
+        gcloud beta pubsub subscriptions create figgy-production-fixity-status --topic figgy-production-fixity-status --expiration-period=never
+        gcloud beta pubsub subscriptions set-iam-policy projects/pulibrary-figgy-storage-1/subscriptions/figgy-production-fixity-status permissions.json
+        rm permissions.json
+        ```
+      - Staging Status Topic is configured with the following commands:
+        ```
+        gcloud beta pubsub topics create figgy-staging-fixity-status
+        echo '{"bindings":[{"members":["serviceAccount:figgy-staging@pulibrary-figgy-storage-1.iam.gserviceaccount.com"],"role":"roles/pubsub.editor"}],"etag":"ACAB"}' > permissions.json
+        gcloud beta pubsub topics set-iam-policy projects/pulibrary-figgy-storage-1/topics/figgy-staging-fixity-status permissions.json
+        gcloud beta pubsub subscriptions create figgy-staging-fixity-status --topic figgy-staging-fixity-status --expiration-period=never
+        gcloud beta pubsub subscriptions set-iam-policy projects/pulibrary-figgy-storage-1/subscriptions/figgy-staging-fixity-status permissions.json
+        rm permissions.json
+        ```
+   1. A daemon will run on each Figgy worker machine which pulls these events
+      and sends them to Sidekiq. This worker can be executed via
+      `RAILS_ENV=production bundle exec rake fixity:run_worker`, but does not
+      manually need to be started - it is handled by a systemd service deployed
+      via Ansible.
+   1. Sidekiq will create `Event` objects for each status sent, which will be
+      displayed in the Fixity Dashboard.
+
+## Consequences
+
+1. Cost
+   1. Retrieval of resources from Google Coldline has a hefty cost associated
+      with it. If we find that 10% per year is too expensive for the benefits of
+      manual fixity checking we may want to scale that down.
```

Comment:
I'm not sure "manual" is the right word here.  But I can see the need to contrast with internal-to-Google fixity checking.  Maybe "calculating our own checksums to verify fixity"?